File: src/main/java/main/tut10/vertexPointLighting/VertexPointLighting.java
///*
// * To change this template, choose Tools | Templates
// * and open the template in the editor.
// */
//package main.tut10.vertexPointLighting;
//
//import com.jogamp.newt.event.KeyEvent;
//import com.jogamp.newt.event.MouseEvent;
//import static com.jogamp.opengl.GL.GL_BACK;
//import static com.jogamp.opengl.GL.GL_CULL_FACE;
//import static com.jogamp.opengl.GL.GL_CW;
//import static com.jogamp.opengl.GL.GL_DEPTH_TEST;
//import static com.jogamp.opengl.GL.GL_DYNAMIC_DRAW;
//import static com.jogamp.opengl.GL.GL_LEQUAL;
//import static com.jogamp.opengl.GL2ES3.GL_COLOR;
//import static com.jogamp.opengl.GL2ES3.GL_DEPTH;
//import static com.jogamp.opengl.GL2ES3.GL_UNIFORM_BUFFER;
//import com.jogamp.opengl.GL3;
//import static com.jogamp.opengl.GL3.GL_DEPTH_CLAMP;
//import com.jogamp.opengl.util.GLBuffers;
//import glutil.BufferUtils;
//import main.framework.Framework;
//import main.framework.Semantic;
//import main.framework.component.Mesh;
//import glm.mat._4.Mat4;
//import glm.quat.Quat;
//import glm.vec._3.Vec3;
//import glm.vec._4.Vec4;
//import glutil.MatrixStack;
//import glutil.Timer;
//import java.io.IOException;
//import java.nio.IntBuffer;
//import java.util.logging.Level;
//import java.util.logging.Logger;
//import javax.xml.parsers.ParserConfigurationException;
//import org.xml.sax.SAXException;
//import view.ObjectData;
//import view.ObjectPole;
//import view.ViewData;
//import view.ViewPole;
//import view.ViewScale;
//
///**
// *
// * @author gbarbieri
// */
//public class VertexPointLighting extends Framework {
//
//    private final String SHADERS_ROOT = "/tut10/vertexPointLighting/shaders",
//            POS_VERTEX_LIGHTING_PN_SHADER_SRC = "pos-vertex-lighting-pn",
//            POS_VERTEX_LIGHTING_PCN_SHADER_SRC = "pos-vertex-lighting-pcn",
//            POS_TRANSFORM_SHADER_SRC = "pos-transform", COLOR_PASSTHROUGH_SHADER_SRC = "color-passthrough",
//            UNIFORM_COLOR_SHADER_SRC = "uniform-color", MESHES_ROOT = "/tut10/data/",
//            CYLINDER_MESH_SRC = "UnitCylinder.xml", PLANE_MESH_SRC = "LargePlane.xml", CUBE_MESH_SRC = "UnitCube.xml";
//
//    public static void main(String[] args) {
//        new VertexPointLighting("Tutorial 10 - Vertex Point Lighting");
//    }
//
//    private ProgramData whiteDiffuseColor, vertexDiffuseColor;
//    private UnlitProgData unlit;
//
//    private Mesh cylinder, plane, cube;
//
//    private ViewData initialViewData = new ViewData(
//            new Vec3(0.0f, 0.5f, 0.0f),
//            new Quat(0.92387953f, 0.3826834f, 0.0f, 0.0f),
//            5.0f,
//            0.0f);
//    private ViewScale viewScale = new ViewScale(
//            3.0f, 20.0f,
//            1.5f, 0.5f,
//            0.0f, 0.0f, //No camera movement.
//            90.0f / 250.0f);
//    private ObjectData initialObjectData = new ObjectData(
//            new Vec3(0.0f, 0.5f, 0.0f),
//            new Quat(1.0f, 0.0f, 0.0f, 0.0f));
//
//    private ViewPole viewPole = new ViewPole(initialViewData, viewScale, MouseEvent.BUTTON1);
//    private ObjectPole objectPole = new ObjectPole(initialObjectData, 90.0f / 250.0f, MouseEvent.BUTTON3, viewPole);
//
//    private IntBuffer projectionUniformBuffer = GLBuffers.newDirectIntBuffer(1);
//
//    private boolean drawColoredCyl = false, drawLight = false;
//    private float lightHeight = 1.5f, lightRadius = 1.0f;
//    private Timer lightTimer = new Timer(Timer.Type.LOOP, 5.0f);
//
//    public VertexPointLighting(String title) {
//        super(title);
//    }
//
//    @Override
//    public void init(GL3 gl3) {
//
//        initializePrograms(gl3);
//
//        try {
//            cylinder = new Mesh(MESHES_ROOT + CYLINDER_MESH_SRC, gl3);
//            plane = new Mesh(MESHES_ROOT + PLANE_MESH_SRC, gl3);
//            cube = new Mesh(MESHES_ROOT + CUBE_MESH_SRC, gl3);
//        } catch (ParserConfigurationException | SAXException | IOException ex) {
//            Logger.getLogger(VertexPointLighting.class.getName()).log(Level.SEVERE, null, ex);
//        }
//
//        gl3.glEnable(GL_CULL_FACE);
//        gl3.glCullFace(GL_BACK);
//        gl3.glFrontFace(GL_CW);
//
//        gl3.glEnable(GL_DEPTH_TEST);
//        gl3.glDepthMask(true);
//        gl3.glDepthFunc(GL_LEQUAL);
//        gl3.glDepthRangef(0.0f, 1.0f);
//        gl3.glEnable(GL_DEPTH_CLAMP);
//
//        gl3.glGenBuffers(1, projectionUniformBuffer);
//        gl3.glBindBuffer(GL_UNIFORM_BUFFER, projectionUniformBuffer.get(0));
//        gl3.glBufferData(GL_UNIFORM_BUFFER, Mat4.SIZE, null, GL_DYNAMIC_DRAW);
//
//        //Bind the static buffers.
//        gl3.glBindBufferRange(GL_UNIFORM_BUFFER, Semantic.Uniform.PROJECTION, projectionUniformBuffer.get(0),
//                0, Mat4.SIZE);
//
//        gl3.glBindBuffer(GL_UNIFORM_BUFFER, 0);
//    }
//
//    private void initializePrograms(GL3 gl3) {
//        whiteDiffuseColor = new ProgramData(gl3, SHADERS_ROOT, POS_VERTEX_LIGHTING_PN_SHADER_SRC, COLOR_PASSTHROUGH_SHADER_SRC);
//        vertexDiffuseColor = new ProgramData(gl3, SHADERS_ROOT, POS_VERTEX_LIGHTING_PCN_SHADER_SRC, COLOR_PASSTHROUGH_SHADER_SRC);
//        unlit = new UnlitProgData(gl3, SHADERS_ROOT, POS_TRANSFORM_SHADER_SRC, UNIFORM_COLOR_SHADER_SRC);
//    }
//
//    @Override
//    public void display(GL3 gl3) {
//
//        gl3.glClearBufferfv(GL_COLOR, 0, clearColor.put(0, 0.0f).put(1, 0.0f).put(2, 0.0f).put(3, 0.0f));
//        gl3.glClearBufferfv(GL_DEPTH, 0, clearDepth.put(0, 1.0f));
//
//        lightTimer.update();
//
//        MatrixStack modelMatrix = new MatrixStack();
//        modelMatrix.setMatrix(viewPole.calcMatrix());
//
//        Vec4 worldLightPosition = calcLightPosition();
//
//        Vec4 lightPosCameraSpace = modelMatrix.top().mul_(worldLightPosition);
//
//        gl3.glUseProgram(whiteDiffuseColor.theProgram);
//        gl3.glUniform3fv(whiteDiffuseColor.lightPosUnif, 1, lightPosCameraSpace.toDfb(vecBuffer));
//        gl3.glUseProgram(vertexDiffuseColor.theProgram);
//        gl3.glUniform3fv(vertexDiffuseColor.lightPosUnif, 1, lightPosCameraSpace.toDfb(vecBuffer));
//
//        gl3.glUseProgram(whiteDiffuseColor.theProgram);
//        gl3.glUniform4f(whiteDiffuseColor.lightIntensityUnif, 0.8f, 0.8f, 0.8f, 1.0f);
//        gl3.glUniform4f(whiteDiffuseColor.ambientIntensityUnif, 0.2f, 0.2f, 0.2f, 1.0f);
//        gl3.glUseProgram(vertexDiffuseColor.theProgram);
//        gl3.glUniform4f(vertexDiffuseColor.lightIntensityUnif, 0.8f, 0.8f, 0.8f, 1.0f);
//        gl3.glUniform4f(vertexDiffuseColor.ambientIntensityUnif, 0.2f, 0.2f, 0.2f, 1.0f);
//        gl3.glUseProgram(0);
//
//        {
//            modelMatrix.push();
//
//            //Render the ground plane.
//            {
//                modelMatrix.push().top().toDfb(matBuffer);
//
//                gl3.glUseProgram(whiteDiffuseColor.theProgram);
//                gl3.glUniformMatrix4fv(whiteDiffuseColor.modelToCameraMatrixUnif, 1, false, matBuffer);
//                modelMatrix.top().toMat3_().toDfb(matBuffer);
//                gl3.glUniformMatrix3fv(whiteDiffuseColor.normalModelToCameraMatrixUnif, 1, false, matBuffer);
//                plane.render(gl3);
//                gl3.glUseProgram(0);
//
//                modelMatrix.pop();
//            }
//            modelMatrix.pop();
//
//            //Render the Cylinder
//            {
//                modelMatrix.push().applyMatrix(objectPole.calcMatrix()).top().toDfb(matBuffer);
//
//                if (drawColoredCyl) {
//                    gl3.glUseProgram(vertexDiffuseColor.theProgram);
//                    gl3.glUniformMatrix4fv(vertexDiffuseColor.modelToCameraMatrixUnif, 1, false, matBuffer);
//                    modelMatrix.top().toMat3_().toDfb(matBuffer);
//                    gl3.glUniformMatrix3fv(vertexDiffuseColor.normalModelToCameraMatrixUnif, 1, false, matBuffer);
//                    cylinder.render(gl3, "lit-color");
//                } else {
//                    gl3.glUseProgram(whiteDiffuseColor.theProgram);
//                    gl3.glUniformMatrix4fv(whiteDiffuseColor.modelToCameraMatrixUnif, 1, false, matBuffer);
//                    modelMatrix.top().toMat3_().toDfb(matBuffer);
//                    gl3.glUniformMatrix3fv(whiteDiffuseColor.normalModelToCameraMatrixUnif, 1, false, matBuffer);
//                    cylinder.render(gl3, "lit");
//                }
//                gl3.glUseProgram(0);
//
//                modelMatrix.pop();
//            }
//
//            //Render the light
//            if (drawLight) {
//
//                modelMatrix.push().translate(worldLightPosition).scale(0.1f, 0.1f, 0.1f);
//
//                gl3.glUseProgram(unlit.theProgram);
//                gl3.glUniformMatrix4fv(unlit.modelToCameraMatrixUnif, 1, false, modelMatrix.top().toDfb(matBuffer));
//                gl3.glUniform4f(unlit.objectColorUnif, 0.8078f, 0.8706f, 0.9922f, 1.0f);
//                cube.render(gl3, "flat");
//
//                modelMatrix.pop();
//            }
//        }
//    }
//
//    private Vec4 calcLightPosition() {
//
//        float currentTimeThroughLoop = lightTimer.getAlpha();
//
//        Vec4 ret = new Vec4(0.0f, lightHeight, 0.0f, 1.0f);
//
//        ret.x = (float) (Math.cos(currentTimeThroughLoop * (Math.PI * 2.0f)) * lightRadius);
//        ret.z = (float) (Math.sin(currentTimeThroughLoop * (Math.PI * 2.0f)) * lightRadius);
//
//        return ret;
//    }
//
//    @Override
//    public void reshape(GL3 gl3, int w, int h) {
//
//        float zNear = 1.0f, zFar = 1_000f;
//        MatrixStack perspMatrix = new MatrixStack();
//
//        perspMatrix.perspective(45.0f, (float) w / h, zNear, zFar);
//
//        gl3.glBindBuffer(GL_UNIFORM_BUFFER, projectionUniformBuffer.get(0));
//        gl3.glBufferSubData(GL_UNIFORM_BUFFER, 0, Mat4.SIZE, perspMatrix.top().toDfb(matBuffer));
//        gl3.glBindBuffer(GL_UNIFORM_BUFFER, 0);
//
//        gl3.glViewport(0, 0, w, h);
//    }
//
//    @Override
//    public void mousePressed(MouseEvent e) {
//        viewPole.mousePressed(e);
//        objectPole.mousePressed(e);
//    }
//
//    @Override
//    public void mouseDragged(MouseEvent e) {
//        viewPole.mouseMove(e);
//        objectPole.mouseMove(e);
//    }
//
//    @Override
//    public void mouseReleased(MouseEvent e) {
//        viewPole.mouseReleased(e);
//        objectPole.mouseReleased(e);
//    }
//
//    @Override
//    public void mouseWheelMoved(MouseEvent e) {
//        viewPole.mouseWheel(e);
//    }
//
//    @Override
//    public void keyPressed(KeyEvent e) {
//
//        switch (e.getKeyCode()) {
//
//            case KeyEvent.VK_ESCAPE:
//                animator.remove(glWindow);
//                glWindow.destroy();
//                break;
//
//            case KeyEvent.VK_SPACE:
//                drawColoredCyl = !drawColoredCyl;
//                break;
//
//            case KeyEvent.VK_I:
//                lightHeight += e.isShiftDown() ? 0.05f : 0.2f;
//                break;
//            case KeyEvent.VK_K:
//                lightHeight -= e.isShiftDown() ? 0.05f : 0.2f;
//                break;
//            case KeyEvent.VK_L:
//                lightRadius += e.isShiftDown() ? 0.05f : 0.2f;
//                break;
//            case KeyEvent.VK_J:
//                lightRadius -= e.isShiftDown() ? 0.05f : 0.2f;
//                break;
//
//            case KeyEvent.VK_Y:
//                drawLight = !drawLight;
//                break;
//
//            case KeyEvent.VK_B:
//                lightTimer.togglePause();
//                break;
//        }
//        if (lightRadius < 0.2f) {
//            lightRadius = 0.2f;
//        }
//    }
//
//    @Override
//    public void end(GL3 gl3) {
//
//        gl3.glDeleteProgram(vertexDiffuseColor.theProgram);
//        gl3.glDeleteProgram(whiteDiffuseColor.theProgram);
//        gl3.glDeleteProgram(unlit.theProgram);
//
//        gl3.glDeleteBuffers(1, projectionUniformBuffer);
//
//        cylinder.dispose(gl3);
//        plane.dispose(gl3);
//        cube.dispose(gl3);
//
//        BufferUtils.destroyDirectBuffer(projectionUniformBuffer);
//    }
//}

File: src/main/kotlin/glNext/tut10/vertexPointLighting.kt
package glNext.tut10

import com.jogamp.newt.event.KeyEvent
import com.jogamp.newt.event.MouseEvent
import com.jogamp.opengl.GL
import com.jogamp.opengl.GL2ES3.*
import com.jogamp.opengl.GL3
import glNext.*
import glm.f
import glm.vec._3.Vec3
import glm.vec._4.Vec4
import glm.quat.Quat
import glm.mat.Mat4
import main.framework.Framework
import main.framework.Semantic
import main.framework.component.Mesh
import uno.buffer.intBufferBig
import uno.glm.MatrixStack
import uno.mousePole.*
import uno.time.Timer
import glm.glm
import uno.buffer.destroy
import uno.glsl.programOf

/**
 * Created by GBarbieri on 23.03.2017.
 */

fun main(args: Array<String>) {
    VertexPointLighting_Next().setup("Tutorial 10 - Vertex Point Lighting")
}

class VertexPointLighting_Next() : Framework() {

    lateinit var whiteDiffuseColor: ProgramData
    lateinit var vertexDiffuseColor: ProgramData
    lateinit var unlit: UnlitProgData

    lateinit var cylinder: Mesh
    lateinit var plane: Mesh
    lateinit var cube: Mesh

    val initialViewData = ViewData(
            Vec3(0.0f, 0.5f, 0.0f),
            Quat(0.92387953f, 0.3826834f, 0.0f, 0.0f),
            5.0f,
            0.0f)
    val viewScale = ViewScale(
            3.0f, 20.0f,
            1.5f, 0.5f,
            0.0f, 0.0f, //No camera movement.
            90.0f / 250.0f)
    val initialObjectData = ObjectData(
            Vec3(0.0f, 0.5f, 0.0f),
            Quat(1.0f, 0.0f, 0.0f, 0.0f))

    val viewPole = ViewPole(initialViewData, viewScale, MouseEvent.BUTTON1)
    val objectPole = ObjectPole(initialObjectData, 90.0f / 250.0f, MouseEvent.BUTTON3, viewPole)

    val projectionUniformBuffer = intBufferBig(1)

    var drawColoredCyl = false
    var drawLight = false
    var lightHeight = 1.5f
    var lightRadius = 1.0f
    val lightTimer = Timer(Timer.Type.Loop, 5.0f)

    override fun init(gl: GL3) = with(gl) {

        initializePrograms(gl)

        cylinder = Mesh(gl, javaClass, "tut10/UnitCylinder.xml")
        plane = Mesh(gl, javaClass, "tut10/LargePlane.xml")
        cube = Mesh(gl, javaClass, "tut10/UnitCube.xml")

        cullFace {
            enable()
            cullFace = back
            frontFace = cw
        }

        depth {
            test = true
            mask = true
            func = lEqual
            range = 0.0 .. 1.0
            clamp = true
        }

        initUniformBuffer(projectionUniformBuffer) {
            data(Mat4.SIZE, GL.GL_DYNAMIC_DRAW)
            //Bind the static buffers.
            range(Semantic.Uniform.PROJECTION, 0, Mat4.SIZE)
        }
    }

    fun initializePrograms(gl: GL3) {
        whiteDiffuseColor = ProgramData(gl, "tut10", "pos-vertex-lighting-PN.vert", "color-passthrough.frag")
        vertexDiffuseColor = ProgramData(gl, "tut10", "pos-vertex-lighting-PCN.vert", "color-passthrough.frag")
        unlit = UnlitProgData(gl, "tut10", "pos-transform.vert", "uniform-color.frag")
    }

    override fun display(gl: GL3) = with(gl) {

        clear {
            color(0)
            depth()
        }

        lightTimer.update()

        val modelMatrix = MatrixStack()
        modelMatrix.setMatrix(viewPole.calcMatrix())

        val worldLightPosition = calcLightPosition()

        val lightPosCameraSpace = modelMatrix.top() * worldLightPosition

        usingProgram(whiteDiffuseColor.theProgram) {
            glUniform3f(whiteDiffuseColor.lightPosUnif, lightPosCameraSpace)
            name = vertexDiffuseColor.theProgram
            glUniform3f(vertexDiffuseColor.lightPosUnif, lightPosCameraSpace)

            name = whiteDiffuseColor.theProgram
            glUniform4f(whiteDiffuseColor.lightIntensityUnif, 0.8f, 0.8f, 0.8f, 1.0f)
            glUniform4f(whiteDiffuseColor.ambientIntensityUnif, 0.2f, 0.2f, 0.2f, 1.0f)
            name = vertexDiffuseColor.theProgram
            glUniform4f(vertexDiffuseColor.lightIntensityUnif, 0.8f, 0.8f, 0.8f, 1.0f)
            glUniform4f(vertexDiffuseColor.ambientIntensityUnif, 0.2f, 0.2f, 0.2f, 1.0f)
        }

        modelMatrix run {

            //Render the ground plane.
            run {
                usingProgram(whiteDiffuseColor.theProgram) {
                    whiteDiffuseColor.modelToCameraMatrixUnif.mat4 = top()
                    val normMatrix = top().toMat3()
                    whiteDiffuseColor.normalModelToCameraMatrixUnif.mat3 = normMatrix
                    plane.render(gl)
                }
            }

            //Render the Cylinder
            run {
                applyMatrix(objectPole.calcMatrix())

                usingProgram {
                    if (drawColoredCyl) {
                        name = vertexDiffuseColor.theProgram
                        vertexDiffuseColor.modelToCameraMatrixUnif.mat4 = top()
                        val normMatrix = top().toMat3()
                        vertexDiffuseColor.normalModelToCameraMatrixUnif.mat3 = normMatrix
                        cylinder.render(gl, "lit-color")
                    } else {
                        name = whiteDiffuseColor.theProgram
                        whiteDiffuseColor.modelToCameraMatrixUnif.mat4 = top()
                        val normMatrix = top().toMat3()
                        whiteDiffuseColor.normalModelToCameraMatrixUnif.mat3 = normMatrix
                        cylinder.render(gl, "lit")
                    }
                }
            }

            //Render the light
            if (drawLight)

                run {
                    translate(worldLightPosition).scale(0.1f)

                    usingProgram(unlit.theProgram) {
                        unlit.modelToCameraMatrixUnif.mat4 = top()
                        glUniform4f(unlit.objectColorUnif, 0.8078f, 0.8706f, 0.9922f, 1.0f)
                        cube.render(gl, "flat")
                    }
                }
        }
    }

    fun calcLightPosition(): Vec4 {

        val currentTimeThroughLoop = lightTimer.getAlpha()

        val ret = Vec4(0.0f, lightHeight, 0.0f, 1.0f)

        ret.x = glm.cos(currentTimeThroughLoop * (glm.PIf * 2.0f)) * lightRadius
        ret.z = glm.sin(currentTimeThroughLoop * (glm.PIf * 2.0f)) * lightRadius

        return ret
    }

    override fun reshape(gl: GL3, w: Int, h: Int) = with(gl) {

        val zNear = 1.0f
        val zFar = 1_000f
        val perspMatrix = MatrixStack()

        perspMatrix.perspective(45.0f, w.f / h, zNear, zFar)

        glBindBuffer(GL_UNIFORM_BUFFER, projectionUniformBuffer)
        glBufferSubData(GL_UNIFORM_BUFFER, perspMatrix.top())
        glBindBuffer(GL_UNIFORM_BUFFER)

        glViewport(w, h)
    }

    override fun mousePressed(e: MouseEvent) {
        viewPole.mousePressed(e)
        objectPole.mousePressed(e)
    }

    override fun mouseDragged(e: MouseEvent) {
        viewPole.mouseDragged(e)
        objectPole.mouseDragged(e)
    }

    override fun mouseReleased(e: MouseEvent) {
        viewPole.mouseReleased(e)
        objectPole.mouseReleased(e)
    }

    override fun mouseWheelMoved(e: MouseEvent) {
        viewPole.mouseWheel(e)
    }

    override fun keyPressed(e: KeyEvent) {

        when (e.keyCode) {

            KeyEvent.VK_ESCAPE -> quit()

            KeyEvent.VK_SPACE -> drawColoredCyl = !drawColoredCyl

            KeyEvent.VK_I -> lightHeight += if (e.isShiftDown) 0.05f else 0.2f
            KeyEvent.VK_K -> lightHeight -= if (e.isShiftDown) 0.05f else 0.2f
            KeyEvent.VK_L -> lightRadius += if (e.isShiftDown) 0.05f else 0.2f
            KeyEvent.VK_J -> lightRadius -= if (e.isShiftDown) 0.05f else 0.2f

            KeyEvent.VK_Y -> drawLight = !drawLight

            KeyEvent.VK_B -> lightTimer.togglePause()
        }
        if (lightRadius < 0.2f)
            lightRadius = 0.2f
    }

    override fun end(gl: GL3) = with(gl) {

        glDeletePrograms(vertexDiffuseColor.theProgram, whiteDiffuseColor.theProgram, unlit.theProgram)

        glDeleteBuffer(projectionUniformBuffer)

        cylinder.dispose(gl)
        plane.dispose(gl)
        cube.dispose(gl)

        projectionUniformBuffer.destroy()
    }

    inner class ProgramData(gl: GL3, root: String, vertex: String, fragment: String) {

        val theProgram = programOf(gl, javaClass, root, vertex, fragment)

        val lightPosUnif = gl.glGetUniformLocation(theProgram, "lightPos")
        val lightIntensityUnif = gl.glGetUniformLocation(theProgram, "lightIntensity")
        val ambientIntensityUnif = gl.glGetUniformLocation(theProgram, "ambientIntensity")

        val modelToCameraMatrixUnif = gl.glGetUniformLocation(theProgram, "modelToCameraMatrix")
        val normalModelToCameraMatrixUnif = gl.glGetUniformLocation(theProgram, "normalModelToCameraMatrix")

        init {
            gl.glUniformBlockBinding(
                    theProgram,
                    gl.glGetUniformBlockIndex(theProgram, "Projection"),
                    Semantic.Uniform.PROJECTION)
        }
    }

    inner class UnlitProgData(gl: GL3, root: String, vertex: String, fragment: String) {

        val theProgram = programOf(gl, javaClass, root, vertex, fragment)

        val objectColorUnif = gl.glGetUniformLocation(theProgram, "objectColor")

        val modelToCameraMatrixUnif = gl.glGetUniformLocation(theProgram, "modelToCameraMatrix")

        init {
            gl.glUniformBlockBinding(
                    theProgram,
                    gl.glGetUniformBlockIndex(theProgram, "Projection"),
                    Semantic.Uniform.PROJECTION)
        }
    }
}